File: 0234.PalindromeLinkedList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func isPalindrome(head *ListNode) bool {
    slow, fast := head, head
    // List will be modified, O(n) + O(1)
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    tail := reverse(slow)
    for head != nil && tail != nil {
        if head.Val != tail.Val {
            return false
        }
        head, tail = head.Next, tail.Next
    }
    /* List will not be modified, O(n) + O(n)
    stack := []int{}
    for fast != nil && fast.Next != nil {
        stack = append(stack, slow.Val)
        slow = slow.Next
        fast = fast.Next.Next
    }
    idx := len(stack) - 1
    if fast != nil { // odd
        slow = slow.Next
    }
    for slow != nil {
        if slow.Val != stack[idx] {
            return false
        }
        slow = slow.Next
        idx--
    }
    */
    return true
}

func reverse(head *ListNode) *ListNode {
    var newHead *ListNode
    for head != nil {
        tmp := head.Next
        head.Next = newHead
        newHead = head
        head = tmp
    }
    return newHead
}

File: 0000-0999/0234.PalindromeLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head is None or head.next is None:
            return True
        fast = slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        sec = None
        while slow:
            p = slow
            slow = slow.next
            p.next = sec
            sec = p
        while head and sec:
            if head.val != sec.val:
                return False
            head = head.next
            sec = sec.next
        return True
        